### PR TITLE
Optimize regex_revalidate.config API endpoint

### LINKED ISSUE #3232: The regex_revalidate.config API endpoint is too slow for large numbers of invalidations
This endpoint can be unbearably slow and CPU-intensive with hundreds of
current invalidation jobs. We should dig into this endpoint (currently still Perl) and see if we can make some easy optimizations.

### PR DESCRIPTION
**Pull request by rawlinp**
#### What does this PR do?

This endpoint can be unbearably slow and CPU-intensive with hundreds of
current invalidation jobs. Do some simple optimization to get this
endpoint more on par with other ATS config file endpoints.

Locally with ~600 current invalidation jobs, I'm seeing about 5-6x speed
improvement with these changes. Because of the added SQL filtering, the
endpoint now returns almost instantly for CDNs that have little or no
current invalidations. Previously, generating this file for ANY CDN
would require iterating over ALL current jobs across ALL CDNs, doing a
bunch of work before throwing away the results from other CDNs.

Fixes #3232

#### Which TC components are affected by this PR?

- [ ] Documentation
- [ ] Grove
- [ ] Traffic Analytics
- [ ] Traffic Monitor
- [x] Traffic Ops
- [ ] Traffic Ops ORT
- [ ] Traffic Portal
- [ ] Traffic Router
- [ ] Traffic Stats
- [ ] Traffic Vault
- [ ] Other _________

#### What is the best way to verify this PR?
1. Create some invalidation jobs for a test DS and compare the before/after output of regex_revalidate.config (with/without this change)
2. Create 500 invalidation jobs and repeat step 1, compare request times

#### Check all that apply

- [ ] This PR includes tests
- [ ] This PR includes documentation updates
- [ ] This PR includes an update to CHANGELOG.md
- [ ] This PR includes all required license headers
- [ ] This PR includes a database migration (ensure that migration sequence is correct)
- [ ] This PR fixes a serious security flaw. Read more: [www.apache.org/security](http://www.apache.org/security/)

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3060/
Test PASSed.

**Comment by rob05c** [comment]
This endpoint is pretty small. Is there a reason not to rewrite it in Go instead?

Go should be even faster, and has the added benefit of reducing load on Postgres compared to using Mojo/workers. Not to mention insisting new changes be in Go will make the rewrite go faster.

**Comment by rawlinp** [comment]
It should be rewritten in Go, but that would take much longer to implement and test than this low-hanging fruit optimization done in the Perl that is only a slight modification of the current implementation. Essentially a SQL filter has been added, and an expensive object initialization has been moved out of a loop -- super simple to test and low-risk to deploy.

**Comment by rawlinp** [comment]
Most of the work done in this endpoint is date parsing and calculations in pure Perl, so I imagine this would get even faster than the optimized Perl simply by being written in Go. The slowness of the Date::Manip Perl package is even a known issue called out in some of the websites I found when investigating it.

**Comment by rawlinp** [comment]
Definitely agree on putting this near the head of the endpoints to port, because it's almost completely open to self-service users who can easily add invalidation jobs to their delivery services.

File: traffic_ops/app/lib/API/Configs/ApacheTrafficServer.pm [hunk 1]
```diff
@@ -1803,15 +1811,11 @@ sub regex_revalidate_dot_config {
...
-		my $job_cdn_id = $row->job_deliveryservice->cdn_id;
-		if ( $cdn_obj->id == $job_cdn_id ) {
-
-			# if there are multipe with same re, pick the longes lasting.
-			if ( !defined( $regex_time{ $row->asset_url } )
-				|| ( defined( $regex_time{ $row->asset_url } ) && $purge_end > $regex_time{ $row->asset_url } ) )
-			{
-				$regex_time{ $row->asset_url } = $purge_end;
-			}
+		# if there are multipe with same re, pick the longes lasting.
```
Comment (dangogh) [inline traffic_ops/app/lib/API/Configs/ApacheTrafficServer.pm]: minor nit -- 2 misspellings

Comment (rawlinp) [reply traffic_ops/app/lib/API/Configs/ApacheTrafficServer.pm]: done

**Comment by mtorluemke** [comment]
thanks @rawlinp !

**Comment by dangogh** [comment]
moving the $date setting outside the loop is obviously correct and a welcome change -- haven't looked at the rest of the changes in detail -- someone signing up for testing this?   definitely +90billion on this...

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3062/
Test PASSed.

File: traffic_ops/app/lib/API/Configs/ApacheTrafficServer.pm (hunk 1 above)
Comment (alficles) [reply traffic_ops/app/lib/API/Configs/ApacheTrafficServer.pm]: Yeah, I saw the spellings. They were there in the original before it was unindented. Probably an error in translation from the original Dutch. 👀

**Pull request merged**